vertx-mongo-client/src/main/java/io/vertx/ext/mongo/impl/MongoClientImpl.java

```diff
@@ -215,23 +224,31 @@ public void close() {
   }
 
   private JsonObject generateIdIfNeeded(JsonObject query, JsonObject update, UpdateOptions options) {
-    if(options.isUpsert() && !update.containsKey(ID_FIELD) && !useObjectId) {
+    if (options.isUpsert() && !update.containsKey(ID_FIELD) && !useObjectId) {
       JsonObject setId = update.getJsonObject("$setOnInsert", new JsonObject());
       String id;
 
-      //This seems odd, but if you filter based on _id, mongo expects the generated _id to match
-      if(query.containsKey(ID_FIELD)) {
+      // This seems odd, but if you filter based on _id, mongo expects the generated _id to match
+      if (query.containsKey(ID_FIELD)) {
         id = query.getString(ID_FIELD);
       } else {
         id = JsonObjectCodec.generateHexObjectId();
       }
       setId.put(ID_FIELD, id);
       update.put("$setOnInsert", setId);
+
+      // If the ID is inside $set, remove it
+      // The error that this would cause would start to get enforced in 3.6 (i.e., $setOnInsert no longer silently
+      // supersedes $set).
+      if (update.containsKey("$set") && update.getJsonObject("$set").containsKey("_id")) {
```

Comment:
not 100% on this one, I do accept that if someone changes the _id to a different value, then it is almost certainly an error by the user, and the mongo will throw an error. However by removing the _id we may be covering up the downstream mistake, and would get into a situation where it looks like we have successfully saved a record from the users point of view but not with the _id that they asked for.

I am a bit 50/50 on what is better here, encode the _id and allow the user to discover their mistake, or remove it.

also this fixes #144

Reply:
Looking at this again, I think it's actually better to keep this exact line in, because it passes the old test and maintains 100% compatibility with the current users (the issues reported) right now.